onion_router/onion_router.py

```diff
@@ -19,21 +21,103 @@
-SOCKET_LISTEN.bind((HOST,PORT))
-SOCKET_LISTEN.listen(1)
-
-try:
-    while 1:
-        conn_socket, client_addr = SOCKET_LISTEN.accept()
-        msg = conn_socket.recv(2048)
-        mod_msg = msg.upper()
-        conn_socket.send(mod_msg)
-        conn_socket.close()
-except KeyboardInterrupt:
-    SOCKET_LISTEN.close()
+LISTENER_SOCKET = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+LISTENER_SOCKET.bind((HOST,PORT))
+LISTENER_SOCKET.listen(50)
+
+
+
+#       MAIN
+########################################################
+
+def main():
+    try:
+        while 1:
+            # accept connection from previous ('backward') node in path
+            back_conn, client_addr = LISTENER_SOCKET.accept()
+            print('Connected to: {}:{}'.format(client_addr[0], str(client_addr[1])))
+
+            # new thread for self-setup as onion_router
+            t = threading.Thread(target=two_way_setup, args=(back_conn,))
+            t.start()
+
+    except (socket.error, KeyboardInterrupt):
+        LISTENER_SOCKET.close()
+
+
+
+#       SETUP TWO TCP CONNECTIONS
+########################################################
+
+# need to establish symkey with back_conn, connect to forw_conn
+def two_way_setup(back_conn):
+    msg = back_conn.recv(2048).decode('utf-8').rstrip()
+
+    # (TODO:decrypt data, initialize+respond with symkey. for now send 'ACK')
+    back_conn.send("ACK".encode('utf-8'))
+
+    # Wait for first-ever onion from back_conn
+    print('[Onion] Waiting for data onion...')
+    msg = back_conn.recv(2048).decode('utf-8').rstrip()
+    print('[Onion] Got data onion.')
+
+    # (TODO:decrypt)
+
+    # parse onion to find out who to send to and what to send
+    msg_dict = json.loads(msg)
+    msg_data = msg_dict["data"]
+    msg_addr = msg_dict["addr"]
+
+    # connect to forw_conn and pass along data from back_conn
+    forw_conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    print('[Onion] Connecting to next onion node...')
+    forw_conn.connect((msg_addr, PORT))
+    print('[Onion] Connected.')
+    forw_conn.send(json.dumps(msg_data).encode('utf-8'))
+
+    # now that two way communication is established, pass data back and forth forever
+
+    t = threading.Thread(target=backward_transfer, args=(forw_conn, back_conn))
+    t.start()
+    forward_transfer(back_conn, forw_conn)
+
+
+
+#       PASS DATA FORWARD AND BACKWARD FOREVER
+########################################################
+
+def forward_transfer(back_conn, forw_conn):
+    while True:
+        msg = back_conn.recv(2048).decode('utf-8').rstrip()
+
+        # (TODO: decrypt)
+
+        # json parse
+        msg_dict = json.loads(msg)
+        msg_data = msg_dict["data"]
+        msg_addr = msg_dict["addr"]
+
+        # pass it on
+        forw_conn.send(json.dumps(msg_data).encode('utf-8'))
+
+def backward_transfer(forw_conn, back_conn):
+    while True:
+        msg = forw_conn.recv(2048).decode('utf-8').rstrip()
+
```

Comment:
To clarify for the testing team: @dliuproduction @Dakos-Industries @JastajVirdee 

The reason why the onion "peels-off" a JSON layer and passes it along in _forward_transfer()_ but simply "passes it along" in _backward_transfer()_ is because:

On its way forward the originator sends a full onion. While on the way backwards (the response) begins as a plain message, that gets encrypted sequentially until it reaches the original originator, who will then apply all the symkeys to obtain the plain response.